### PR TITLE
tstest/natlab/vnet: add start of IPv6 support

### PR DESCRIPTION
**Pull request by bradfitz**
Updates #13038

### CODE REVIEW

**Comment by bradfitz** [comment]
Gets a global IPv6 address now but UDP STUN queries don't yet work. Ran out of time to debug why not so marking draft. Will debug later if nobody beats me to it.

File: tstest/natlab/vnet/vnet.go [hunk 1]
```diff
@@ -144,25 +152,48 @@ func (n *network) initStack() error {
...
-	}
+	var routes []tcpip.Route
 
-	ipv4Subnet, err := tcpip.NewSubnet(tcpip.AddrFromSlice(make([]byte, 4)), tcpip.MaskFromBytes(make([]byte, 4)))
-	if err != nil {
-		return fmt.Errorf("could not create IPv4 subnet: %v", err)
-	}
-	n.ns.SetRouteTable([]tcpip.Route{
-		{
+	if n.lanIP.Addr().IsValid() {
```
Comment (bradfitz) [inline tstest/natlab/vnet/vnet.go:157]: if n.v4

File: tstest/natlab/vnet/vnet.go [hunk 2]
```diff
@@ -144,25 +152,48 @@ func (n *network) initStack() error {
...
+			return fmt.Errorf("could not create IPv4 subnet: %v", err)
+		}
+		routes = append(routes, tcpip.Route{
 			Destination: ipv4Subnet,
 			NIC:         nicID,
-		},
-	})
+		})
+	}
+	if n.wanIP6.Addr().IsValid() {
```
Comment (bradfitz) [inline tstest/natlab/vnet/vnet.go:176]: if n.v6

File: tstest/natlab/vnet/vnet.go [hunk 3]
```diff
@@ -183,22 +214,39 @@ func (n *network) initStack() error {
 			}
```
Comment (bradfitz) [inline tstest/natlab/vnet/vnet.go:214]: line 205: pull this out into its own func, not big closure

File: tstest/natlab/vnet/vnet.go [hunk 4]
```diff
@@ -277,7 +337,7 @@ func (n *network) acceptTCP(r *tcp.ForwarderRequest) {
 	}
 
 	if destPort == 124 {
-		node, ok := n.nodesByIP[clientRemoteIP]
+		node, ok := n.nodeForDestIP(clientRemoteIP) //nodesByIP[clientRemoteIP]
```
Comment (bradfitz) [inline tstest/natlab/vnet/vnet.go:340]: clean

File: tstest/natlab/vnet/vnet.go [hunk 5]
```diff
@@ -1102,7 +1302,12 @@ func (n *network) HandleEthernetIPv4PacketForRouter(ep EthernetPacket) {
 			InterfaceIndex: n.lanInterfaceID,
 		}, buf)
 
-		src = n.doNATOut(src, dst)
+		wanSrc := n.doNATOut(src, dst)
+		if !wanSrc.IsValid() {
+			n.logf("warning: NAT dropped packet; no NAT out mapping for %v=>%v", src, dst)
+			return
+		}
```
Comment (bradfitz) [inline tstest/natlab/vnet/vnet.go:1309]: send this separately

File: tstest/integration/nat/nat_test.go [hunk 6]
```diff
@@ -445,6 +453,12 @@ func TestEasyEasy(t *testing.T) {
 	nt.want(routeDirect)
 }
 
+func TestJustIPv6(t *testing.T) {
```
Comment (maisem) [inline tstest/integration/nat/nat_test.go:456]: What about 4+6 on one side and Just6 on the other

File: tstest/natlab/vnet/conf.go [hunk 7]
```diff
@@ -141,7 +142,11 @@ func (c *Config) AddNetwork(opts ...any) *Network {
 			if ip, err := netip.ParseAddr(o); err == nil {
 				n.wanIP = ip
 			} else if ip, err := netip.ParsePrefix(o); err == nil {
-				n.lanIP = ip
+				if ip.Addr().Is4() {
+					n.lanIP = ip
+				} else if ip.Addr().Is6() {
+					n.wanIP6 = ip
```
Comment (maisem) [inline tstest/natlab/vnet/conf.go:148]: Hmm, leave a comment here that the wantIP6 is deliberate and not a typo. The if block has lanIP

File: tstest/natlab/vnet/conf.go [hunk 8]
```diff
@@ -333,10 +383,13 @@ func (s *Server) initFromConfig(c *Config) error {
 		// Allocate a lanIP for the node. Use the network's CIDR and use final
```
Comment (maisem) [inline tstest/natlab/vnet/conf.go:383]: Move inside if?

File: tstest/natlab/vnet/vnet.go [hunk 9]
```diff
@@ -832,34 +886,55 @@ func (s *Server) routeUDPPacket(up UDPPacket) {
...
 		n.logf("dropping write of packet from %v to itself", srcMAC)
-		return
+		return false
 	}
 	if nw, ok := n.writers.Load(dstMAC); ok {
 		nw.write(res)
-		return
+		return true
 	}
+	// gp := gopacket.NewPacket(res, layers.LayerTypeEthernet, gopacket.Lazy)
```
Comment (maisem) [inline tstest/natlab/vnet/vnet.go:914]: Remove

File: tstest/natlab/vnet/vnet.go [hunk 10]
```diff
@@ -874,20 +949,45 @@ func (n *network) HandleEthernetPacket(ep EthernetPacket) {
...
+		if dstMAC == macAllRouters {
+			if rs, ok := ep.gp.Layer(layers.LayerTypeICMPv6RouterSolicitation).(*layers.ICMPv6RouterSolicitation); ok {
+				n.handleIPv6RouterSolicitation(ep, rs)
+			} else {
+				n.logf("unexpected IPv6 packet to all-routers: %v", ep.gp)
+			}
+			return
+		}
+		if dstMAC[0] == 0x33 && dstMAC[1] == 0x33 {
+			if ns, ok := ep.gp.Layer(layers.LayerTypeICMPv6NeighborSolicitation).(*layers.ICMPv6NeighborSolicitation); ok {
```
Comment (maisem) [inline tstest/natlab/vnet/vnet.go:965]: Comment here about 0x3333 being special

File: tstest/natlab/vnet/vnet.go [hunk 11]
```diff
@@ -874,20 +949,45 @@ func (n *network) HandleEthernetPacket(ep EthernetPacket) {
...
+				n.handleIPv6NeighborSolicitation(ep, ns)
+			} else {
+				if ep.gp.Layer(layers.LayerTypeMLDv2MulticastListenerReport) != nil {
+					return
```
Comment (maisem) [inline tstest/natlab/vnet/vnet.go:969]: Comment here why

**Review by maisem — APPROVED** [review]

File: tstest/natlab/vnet/conf.go (hunk 7 above)
Comment (bradfitz) [reply tstest/natlab/vnet/conf.go:148]: done, and renamed some fields

File: tstest/natlab/vnet/vnet.go (hunk 10 above)
Comment (bradfitz) [reply tstest/natlab/vnet/vnet.go:965]: better:

```go
// IsIPv6Multicast reports whether m is an IPv6 multicast MAC address,
// typically one containing a solicited-node multicast address.
func (m MAC) IsIPv6Multicast() bool {
	return m[0] == 0x33 && m[1] == 0x33
}
```

File: tstest/integration/nat/nat_test.go (hunk 6 above)
Comment (bradfitz) [reply tstest/integration/nat/nat_test.go:456]: Done. But these tests don't quite work yet for reasons I haven't yet debugged.

IPv6 works by hand in a gokrazy shell.

**Comment by bradfitz** [comment]
Still buggy/incomplete, but good enough checkpoint.

**Pull request merged**